tests/page-objects/default/contacts/contacts.wdio.page.js

```diff
@@ -5,72 +5,76 @@ const utils = require('@utils');
 const modalPage = require('@page-objects/default/common/modal.wdio.page');
 const searchPage = require('@page-objects/default/search/search.wdio.page');
 const mobileSearchPage = require('@page-objects/default-mobile/search/search.wdio.page');
+const commonEnketoPage = require('@page-objects/default/enketo/common-enketo.wdio.page');
 
+// search
 const searchBox = () => $('.mm-search-bar-container input#freetext');
-const contentRowSelector = '#contacts-list .content-row';
+
+// menu
+const menuContent = (testId) => $(`.mat-mdc-menu-content .mat-mdc-menu-item[test-id="${testId}"]`);
+const exportButton = () => menuContent('export-contacts');
```

Comment:
If I use `{}` it will force me to put the `return` in a new line

![image](https://github.com/medic/cht-core/assets/94494491/92304299-ecb8-40f8-a91c-cbfdc4fd5724)

And it would look like this:

![image](https://github.com/medic/cht-core/assets/94494491/e54da33d-da5f-49c3-8a13-27f257b45d0e)

Is it incorrect if I leave it as it was? 🤔 

![image](https://github.com/medic/cht-core/assets/94494491/e19881f4-1a22-49a2-a8f9-b255499545de)

Reply:
Nevermind 😅 everything failed and I found why....

> _Arrow functions do not have their own `this` context; they inherit `this` from the parent scope at the time they are defined. This means that `this` inside an arrow function refers to the `this` value of the enclosing non-arrow function or the global object if there is no enclosing function._

Reply:
> If I use {} it will force me to put the return in a new line

I see. Well, in my opinion, the benefit of keeping it organized is greater than keeping everything in one line. Another less cool version is keeping constants outside the scoping object. 

```

const taskCardElement = {
  TASK_CONTAINER: '.card.tasks'
   container() { // <-- it needs to be this kind of fn, not arrow fn. 
      return $(this.TASK_CONTAINER);
   }
   filter() {
      return $(this.TASK_CONTAINER + '.table-filter a');
   }
}


VS constants outside obj:

const TASK_CONTAINER = '.card.tasks'; // keep them close to scoping object.
const taskCardElement = {
   container: () => $(this.TASK_CONTAINER);
   filter: () => $(this.TASK_CONTAINER + '.table-filter a')
}
```

Reply:
A separate note on these constants:
![image](https://github.com/medic/cht-core/assets/94494491/e54da33d-da5f-49c3-8a13-27f257b45d0e)


Consider chaining selectors to reduce having too many constants, for example: 
```javascript
const taskCardElement = {
  TASK_CONTAINER: '.card.tasks'
   container() { 
      return $(this.TASK_CONTAINER);
   }
   filter() {
      return $(this.TASK_CONTAINER + '.table-filter'); <-- append strings 
   }
   firstFilter() {
      return  this.filter().$('a:first-child()'); <--- chaining to reuse, no need constant
   }
}
```

Disclaimer: I'm not sure if chaining is slower in performance, you might need to check.

Reply:
Ok, new update about this thing 😅 

There is one method, that receives as a parameter the function that I am creating in the object that returns a selector, the problem is that the function uses `this.` to reference another function that returns a string, and the `this` is not bound to the correct context so the test is failing because it is saying that the function that I am calling is not a function. 

I know that this is a little confusing so let me try it with an example:

we have this object defined:

```.js
const leftPanelSelectors = {
  CONTACT_LIST_SELECTOR: '#contacts-list',
  CONTENT_ROW_SELECTOR() {
    const test = `${this.CONTACT_LIST_SELECTOR} .content-row`;
    return test;
  },
  contactList() {
    return $(this.CONTACT_LIST_SELECTOR);
  },
  contentRow() {
    return $(this.CONTENT_ROW_SELECTOR());
  },
  contentRows() {
    return $$(this.CONTENT_ROW_SELECTOR());
  },
  contactName() {
    return $$(`${this.CONTENT_ROW_SELECTOR()} .heading h4 span`);
  },
  contactListLoadingStatus() {
    return $(`${this.CONTACT_LIST_SELECTOR} .loading-status`);
  },
};
```

and in this method we are calling `getTextForElements`

```.js
const getAllLHSContactsNames = async () => {
  await (await leftPanelSelectors.contentRow()).waitForDisplayed();
  return commonPage.getTextForElements(leftPanelSelectors.contactName);
};
```

the method `getTextForElements` will iterate through the `contactName` elements

```.js
const getTextForElements = async (elements) => {
  const elems = await elements();
  return elems.map(elem => elem.getText());
};
```

that is when it fails, saying that `this.CONTENT_ROW_SELECTOR is not a function`.

I was talking with @jkuester and he solved the problem using the method `.bind` but we think that it will be really confusing to do that, in all the different places. The method will look like this:

```.js
const getAllLHSContactsNames = async () => {
  await (await leftPanelSelectors.contentRow()).waitForDisplayed();
  const { contactName } = leftPanelSelectors;
  const test = contactName.bind(leftPanelSelectors);
  return commonPage.getTextForElements(test);
};
```

So going back to my previous comment, I think that the easiest solution will be to go with **option 1**

Reply:
Yes, I agree, that's confusing. 

You can consider using selector chaining to reduce the amount of constants outside the object.[ Like the example I mentioned yesterday.](https://github.com/medic/cht-core/pull/9221#discussion_r1668389234)

Reply:
IMO - I prefer option 2 where we have the regular functions as it looks clearer (although taking more lines) but since it's usage introduces the extra complexity of adding the `bind()` method, it's no longer fist choice - unless there's a better way to implement this and avoid the `ERROR TypeError:`

Reply:
> You can consider using selector chaining to reduce the amount of constants outside the object

The problem with doing that is that we will need to use `this` to access it, and that will bring us problems again 😅 we will need to move it outside of the scoping object and it will be the same as having a constant 🤷‍♀️